Feature/add autonomous agent classes

ahmadhp opened the pull request:
Feature: https://github.com/ahmadhp/web3/issues/1

- [x] Created an autonomous agent system
- [x] Continuously consume messages (of different types) from an InBox
- [x] Emit messages to an OutBox
- [x] Allow for registration of message handlers to handle a given message type with its
- [x] Allow for registration of behaviours
- [x] unit and integration tests

ahmadhp committed:
Added Agent, orchestrator, Readme and unit tests

ahmadhp committed:
issues/1 - added integration test and gitignore

ahmadhp committed:
issues/1 - updated readme

ahmadhp committed:
issues/1 - updated readme

ahmadhp committed:
issues/1 - updated doc

ahmadhp committed:
issues/1 - tidy up tests

angrybayblade commented:
@ahmadhp I get following error when I run the unit tests

```
virajpatel@Virajs-MacBook-Pro web3 % python3 -m unittest discover tests
.Task exception was never retrieved
future: <Task finished name='Task-5' coro=<AutonomousAgent.run_behaviors() done, defined at /Users/virajpatel/valory/assignments/web3/web3/agent.py:46> exception=AttributeError("'AutonomousAgent' object has no attribute 'agent_id'") created at /opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/tasks.py:337>
source_traceback: Object created at (most recent call last):
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/__main__.py", line 18, in <module>
    main(module=None)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/main.py", line 101, in __init__
    self.runTests()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/main.py", line 271, in runTests
    self.result = testRunner.run(self.test)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/runner.py", line 184, in run
    test(result)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/suite.py", line 122, in run
    test(result)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/suite.py", line 122, in run
    test(result)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/suite.py", line 122, in run
    test(result)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 650, in __call__
    return self.run(*args, **kwds)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 159, in run
    return super().run(result)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 64, in _callTestMethod
    self._callMaybeAsync(method)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 87, in _callMaybeAsync
    return self._asyncioTestLoop.run_until_complete(fut)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
    self._run_once()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 1901, in _run_once
    handle._run()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 101, in _asyncioLoopRunner
    ret = await awaitable
  File "/Users/virajpatel/valory/assignments/web3/tests/test_agent.py", line 22, in test_run_behaviors
    _ = asyncio.create_task(self.agent.run_behaviors())
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/tasks.py", line 337, in create_task
    task = loop.create_task(coro)
Traceback (most recent call last):
  File "/Users/virajpatel/valory/assignments/web3/web3/agent.py", line 53, in run_behaviors
    await self.emit_message(message)
  File "/Users/virajpatel/valory/assignments/web3/web3/agent.py", line 70, in emit_message
    print(f"Emitting message to Outbox of Agent {self.agent_id}", message)
AttributeError: 'AutonomousAgent' object has no attribute 'agent_id'
/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py:101: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
Coroutine created at (most recent call last)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 64, in _callTestMethod
    self._callMaybeAsync(method)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 87, in _callMaybeAsync
    return self._asyncioTestLoop.run_until_complete(fut)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
    self._run_once()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 1901, in _run_once
    handle._run()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/virajpatel/valory/assignments/web3/web3/agent.py", line 51, in run_behaviors
    message = behavior()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/mock.py", line 1114, in __call__
    return self._mock_call(*args, **kwargs)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/mock.py", line 1118, in _mock_call
    return self._execute_mock_call(*args, **kwargs)
  ret = await awaitable
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
...Setting up message handler to send outbox message of Agent 1 to inbox of Agent 2
Emitting message to Outbox of Agent 1 {'type': 'random', 'content': 'sky universe'}
Passing message from Outbox of Agent 1 to Inbox of 2
Emitting message to Outbox of Agent 2 {'type': 'random', 'content': 'sky human'}
Message received by agent 2 : {'type': 'random', 'content': 'sky universe'}
removed from inbox 2  0
Emitting message to Outbox of Agent 1 {'type': 'random', 'content': 'human crypto'}
Passing message from Outbox of Agent 1 to Inbox of 2
Emitting message to Outbox of Agent 2 {'type': 'random', 'content': 'space space'}
Message received by agent 2 : {'type': 'random', 'content': 'human crypto'}
removed from inbox 2  0
/Users/virajpatel/valory/assignments/web3/tests/test_agent_communication.py:43: RuntimeWarning: coroutine 'AgentOrchestrator.stop_agents' was never awaited
Coroutine created at (most recent call last)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 159, in run
    return super().run(result)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 64, in _callTestMethod
    self._callMaybeAsync(method)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 87, in _callMaybeAsync
    return self._asyncioTestLoop.run_until_complete(fut)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
    self._run_once()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 1901, in _run_once
    handle._run()
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/opt/homebrew/Cellar/python@3.10/3.10.13_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/async_case.py", line 101, in _asyncioLoopRunner
    ret = await awaitable
  File "/Users/virajpatel/valory/assignments/web3/tests/test_agent_communication.py", line 43, in test_message_passing
    self.orchestrator.stop_agents()
  self.orchestrator.stop_agents()
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
.
----------------------------------------------------------------------
Ran 5 tests in 8.046s

OK
```

The cause of the error seems to be missing class variable

```
  File "/Users/virajpatel/valory/assignments/web3/web3/agent.py", line 70, in emit_message
    print(f"Emitting message to Outbox of Agent {self.agent_id}", message)
AttributeError: 'AutonomousAgent' object has no attribute 'agent_id'
```

ahmadhp commented:
@angrybayblade Test works fine if we run via PyCharm IDE. let me investigate the command issue, please.

ahmadhp committed:
issues/1 - fixed warnings in the tests

ahmadhp commented:
@angrybayblade The tests warnings are fixed now. Thanks for reporting :bow: 

- python3 -m unittest discover tests 
- Emitting message to Outbox of Agent agent1 {'type': 'random', 'content': 'ocean world'}
- Emitting message to Outbox of Agent agent1 <Future finished result={'content': 'Test behavior', 'type': 'test'} created at /home/ahmad/code/web3/tests/test_agent.py:23>
- Emitting message to Outbox of Agent agent1 {'type': 'random', 'content': 'ocean crypto'}
- Emitting message to Outbox of Agent agent1 <Future finished result={'content': 'Test behavior', 'type': 'test'} created at /home/ahmad/code/web3/tests/test_agent.py:23>
- Emitting message to Outbox of Agent agent1 {'type': 'random', 'content': 'ocean moon'}
- Emitting message to Outbox of Agent agent1 <Future finished result={'content': 'Test behavior', 'type': 'test'} created at /home/ahmad/code/web3/tests/test_agent.py:23>
- ...Setting up message handler to send outbox message of Agent 1 to inbox of Agent 2
- Emitting message to Outbox of Agent 1 {'type': 'random', 'content': 'world moon'}
- Passing message from Outbox of Agent 1 to Inbox of 2
- Emitting message to Outbox of Agent 2 {'type': 'random', 'content': 'universe crypto'}
- Message received by agent 2 : {'type': 'random', 'content': 'world moon'}
- removed from inbox 2  0
- Emitting message to Outbox of Agent 1 {'type': 'random', 'content': 'space space'}
- Passing message from Outbox of Agent 1 to Inbox of 2
- Emitting message to Outbox of Agent 2 {'type': 'random', 'content': 'crypto universe'}
- Message received by agent 2 : {'type': 'random', 'content': 'space space'}
- removed from inbox 2  0

- ----------------------------------------------------------------------
- Ran 5 tests in 8.046s
- 
- OK
